File: leetcode/top-interview-150/121/Solution.java
/**
 * BEST TIME TO BUY AND SELL STOCK
 *
 * You are given an array prices where prices[i] is the price of a given stock
 * on the ith day
 * You want to maximize your profit by choosing a single day to buy one stock
 * and choosing a different day in the future to sell that stock
 * Return the maximum profit you can achieve from this transaction
 * If you cannot achieve any profit, return 0
 */

public class Solution {
  public static int maxProfit(int[] prices) {
    int n = prices.length;
    int min = 0;
    int max = 0;
    int profit = 0;

    for (int i = 0; i < n; i++) {
      if (prices[min] > prices[i]) {
        min = i;
      }
      max = prices[i] - prices[min];
      profit = Math.max(profit, max);
    }

    return profit;
  }

  public static void main(String[] args) {
    int[] prices = { 7, 1, 5, 3, 6, 4 };
    int rs = maxProfit(prices);
    System.out.println(rs);
  }
}

File: leetcode/top-interview-150/121/Solution.go
package main

import (
	"fmt"
)

func main() {
	prices := [6]int{7, 1, 5, 3, 6, 4}
	fmt.Println(maxProfit(prices[:]))
}

func maxProfit(prices []int) int {
	maxProfit := 0
	minPrice := prices[0]

	for _, price := range prices {
		if minPrice > price {
			minPrice = price
		}
		if price-minPrice > maxProfit {
			maxProfit = price - minPrice
		}
	}

	return maxProfit
}
